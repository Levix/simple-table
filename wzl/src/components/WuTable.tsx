import { defineComponent, provide, reactive, ref } from "vue";
import WuHead from "./Head";
import WuBody from "./Body";
import WuFoot from "./Foot";
import { type TableProps, tableProps } from "./types";
import { useDataSource } from "../composables/useDataSource";
import { useColumns } from "../composables/useColumns";
// import { usePagetion } from "../composables/usePagetion";
import { TABLE_TOKEN } from "./types";

export default defineComponent({
  name: "WuTable",
  props: tableProps,
  setup(props: TableProps, { attrs, emit, slots }) {
    /** 处理columns */
    const { handledColumns } = useColumns(props);
    const { sortData } = useDataSource(props, handledColumns.value);

    const sortList: any[] = reactive([]);
    const pageSize = ref(1);

    const context = {
      sortData,
      handledColumns,
    };

    // 排序
    const changeSort = (params: any[]) => {
      sortList.length = 0;
      sortList.push(params[0]);
      sortList.push(params[1]);
      console.log(params, sortList, "🍉");
    };

    // 分页
    const changePage = (params) => {
      pageSize.value = params;
      console.log(params, "🍍");
    };

    provide(TABLE_TOKEN, context);

    return () => {
      return (
        <table
          width="90%"
          border="1"
          cellspacing="0"
          cellpadding="4"
          align="center"
        >
          <WuHead onChangeSort={changeSort} />
          <WuBody sortOptions={sortList} pageOptions={pageSize} />
          <WuFoot onChangePage={changePage} />
        </table>
      );
    };
  },
});
