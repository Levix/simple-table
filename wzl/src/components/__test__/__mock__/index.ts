import type { TableColumns, TableDataSourceType } from '../../types';
import { useDataSource } from "../../../composables/useDataSource";
import { useColumns } from "../../../composables/useColumns";
import { ref } from 'vue';

const dataSource: TableDataSourceType[] = []
for (let index = 0; index < 10; index++) {
    dataSource.push({
        key: index,
        name: `Edrward ${index}`,
        age: 18 + index,
        address: `London Park no. ${index}`,
        tags: ['nice', 'developer'],
    })
}
// 表格源数据
export { dataSource };
// 表格列配置
export const columns: TableColumns[] = [
    {
		title: "Name",
		key: "name",
	},
	{
		title: "Age",
		key: "age",
		width: 100,
		align:'center',
		sortable: {
			orderBy: "DESC",
			sorter: (curr, next) => next - curr,
		},
	},
	{
		title: "Address",
		key: "address",
	},
	{
		title: "Tags",
		key: "tags",
	},
];

export const sortIconObj = {
    ascTriangle: "▲",
    descTriangle: "▼",
    normalTriangle: "■",
}

export const a_columns: TableColumns[] = [
    {
        title: "Col_1",
        key: "Col__1",
    },
    {
        title: "Col_2",
        key: "Col__2",
        sortable: {
            orderBy: "asc"
        }
    },
    {
        title: "Col_3",
        key: "Col__3",
    },
    {
        title: "Col_4",
        key: "Col__4",
    },
]
export const d_columns: TableColumns[] = [
    {
        title: "Col_1",
        key: "Col__1",
    },
    {
        title: "Col_2",
        key: "Col__2",
        sortable: {
            orderBy: "desc"
        }
    },
    {
        title: "Col_3",
        key: "Col__3",
    },
    {
        title: "Col_4",
        key: "Col__4",
    },
]

// 处理后的配置
export const { newColumns } = useColumns(columns);
// 处理后的数据
export const { sortTableData } = useDataSource(ref(dataSource), newColumns);
// body配置
export const bodyOptions = {
    sortData: sortTableData.value,
    columns: newColumns.value,
    sortOption: {
        key: 1,
        sortStatus: 'asc',
    },
    pageSize: 10,
    curPageNumber: 1,
}

// foot配置
export const footOptions = {
    sortData: sortTableData.value,
    pageSize: 4,
}

export const newFootOptions = {
    sortData: sortTableData.value,
    pageSize: 8,
}