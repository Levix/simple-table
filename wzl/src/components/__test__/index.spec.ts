import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import WuTable from '../index';
import { dataSource, columns } from './__mock__/index';

describe('WuTable', () => {
    const wrapper = mount(WuTable, {
        props: {
            dataSource,
            columns,
            pageSize: 10,
        }
    })

    it('打个快照 WuTable', () => {
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.exists()).toBe(true);
    })

    it('组件是否正常渲染', () => {
        /** 表格是否存在 */
        expect(wrapper.find('table').exists()).toBe(true);

        /** 表头是否存在 */
        expect(wrapper.find('thead').exists()).toBe(true);

        /** 内容是否存在 */
        expect(wrapper.find('tbody').exists()).toBe(true);

        /** 底部是否存在 */
        expect(wrapper.find('tfoot').exists()).toBe(true);
    })
})