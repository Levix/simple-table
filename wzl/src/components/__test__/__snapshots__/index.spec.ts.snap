// Vitest Snapshot v1

exports[`WuTable > 打个快照 WuTable 1`] = `
"<table>
  <thead class=\\"wu_head\\">
    <tr>
      <th title=\\"Name\\" class=\\"align__left width__200px\\">Name
        <!---->
      </th>
      <th title=\\"Age\\" class=\\"align__center width__100px\\">Age<span class=\\"wu_head__sort\\">▼</span></th>
      <th title=\\"Address\\" class=\\"align__left width__200px\\">Address
        <!---->
      </th>
      <th title=\\"Tags\\" class=\\"align__left width__200px\\">Tags
        <!---->
      </th>
    </tr>
  </thead>
  <tbody class=\\"wu_body\\">
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 9</td>
      <td class=\\"body_align__center\\">27</td>
      <td class=\\"body_align__left\\">London Park no. 9</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 8</td>
      <td class=\\"body_align__center\\">26</td>
      <td class=\\"body_align__left\\">London Park no. 8</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 7</td>
      <td class=\\"body_align__center\\">25</td>
      <td class=\\"body_align__left\\">London Park no. 7</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 6</td>
      <td class=\\"body_align__center\\">24</td>
      <td class=\\"body_align__left\\">London Park no. 6</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 5</td>
      <td class=\\"body_align__center\\">23</td>
      <td class=\\"body_align__left\\">London Park no. 5</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 4</td>
      <td class=\\"body_align__center\\">22</td>
      <td class=\\"body_align__left\\">London Park no. 4</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 3</td>
      <td class=\\"body_align__center\\">21</td>
      <td class=\\"body_align__left\\">London Park no. 3</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 2</td>
      <td class=\\"body_align__center\\">20</td>
      <td class=\\"body_align__left\\">London Park no. 2</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 1</td>
      <td class=\\"body_align__center\\">19</td>
      <td class=\\"body_align__left\\">London Park no. 1</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
    <tr class=\\"wu_body__tr\\">
      <td class=\\"body_align__left\\">Edrward 0</td>
      <td class=\\"body_align__center\\">18</td>
      <td class=\\"body_align__left\\">London Park no. 0</td>
      <td class=\\"body_align__left\\">nicedeveloper</td>
    </tr>
  </tbody>
  <tfoot class=\\"wu_foot\\">
    <td colspan=\\"4\\"><button class=\\"wu_foot__prev prev_disabled\\" disabled=\\"\\">前一页</button><button class=\\"actived\\">1</button><button class=\\"wu_foot__next next_disabled\\" disabled=\\"\\">后一页</button>
      <div class=\\"wu_foot__box\\"><span>当前：第1页</span><br><span>10条 / 页</span><br><span>总共：10条</span></div>
    </td>
  </tfoot>
</table>"
`;
